File: frontend/src/invoiceSources.ts
import type { Settlement } from "./types";

export function settlementSourcesFollowOriginal(originalIds: number[], replacementIds: number[], settlementById: Map<number, Settlement>, sameSources = false) {
  if (!originalIds.length) return false;
  const originals = originalIds.map((id) => settlementById.get(id));
  if (originals.some((item) => !item)) return false;
  if (sameSources) return originalIds.length === replacementIds.length
    && new Set(replacementIds).size === originalIds.length
    && replacementIds.every((id) => originalIds.includes(id) && settlementById.get(id)?.status === "FINALIZED");
  const original = originals[0]!;
  const sourceGroup = (item: Settlement) => `${item.platform_id}:${item.fee_plan_id}`;
  const originalGroups = new Set(originals.map((item) => sourceGroup(item!)));
  const originalIdSet = new Set(originalIds);
  const matched = new Set<number>();
  const groups = new Set<string>();
  for (const id of replacementIds) {
    let current = settlementById.get(id);
    if (!current || current.status !== "FINALIZED"
      || current.client_id !== original.client_id || current.year !== original.year
      || current.quarter !== original.quarter
      || groups.has(sourceGroup(current))) return false;
    groups.add(sourceGroup(current));
    if (!originalGroups.has(sourceGroup(current))) continue;
    const visited = new Set<number>([current.id]);
    let found = false;
    while (current?.replaces_settlement_id != null) {
      const previousId = current.replaces_settlement_id;
      if (visited.has(previousId)) return false;
      visited.add(previousId);
      if (originalIdSet.has(previousId)) {
        if (matched.has(previousId)) return false;
        matched.add(previousId);
        found = true;
        break;
      }
      current = settlementById.get(previousId);
    }
    if (!found) return false;
  }
  return matched.size === originalIdSet.size;
}

File: backend/app/services/invoice_sources.py
from sqlalchemy.orm import Session

from ..models import Invoice, QuarterlySettlement


def settlements_follow_original(db: Session, original: Invoice, settlements: list[QuarterlySettlement], *, same_sources: bool = False) -> bool:
    """Cover every old source; new platform/plan combinations may join the bill."""
    originals = {source.settlement_id: db.get(QuarterlySettlement, source.settlement_id) for source in original.sources}
    if not originals or any(item is None for item in originals.values()):
        return False
    if same_sources:
        return (
            len(settlements) == len(originals)
            and {item.id for item in settlements} == set(originals)
            and all(item.status == "FINALIZED" for item in settlements)
        )
    original_groups = {(item.platform_id, item.fee_plan_id) for item in originals.values()}
    matched_ids: set[int] = set()
    seen_groups: set[tuple[int, int]] = set()
    group = (original.client_id, original.year, original.quarter)
    for settlement in settlements:
        if (
            settlement.status != "FINALIZED"
            or (settlement.client_id, settlement.year, settlement.quarter) != group
            or (settlement.platform_id, settlement.fee_plan_id) in seen_groups
        ):
            return False
        source_group = (settlement.platform_id, settlement.fee_plan_id)
        seen_groups.add(source_group)
        if source_group not in original_groups:
            continue
        current = settlement
        visited = {current.id}
        while current.replaces_settlement_id is not None:
            previous_id = current.replaces_settlement_id
            if previous_id in visited:
                return False
            visited.add(previous_id)
            if previous_id in originals:
                if previous_id in matched_ids:
                    return False
                matched_ids.add(previous_id)
                break
            current = db.get(QuarterlySettlement, previous_id)
            if current is None:
                return False
        else:
            return False
    return matched_ids == set(originals)
